data/base.go

```diff
@@ -0,0 +1,256 @@
+
+	//optional data
+	DeviceTime       *string       `json:"deviceTime,omitempty" bson:"deviceTime,omitempty" valid:"omitempty,timestr"`
+	TimezoneOffset   *int          `json:"timezoneOffset,omitempty" bson:"timezoneOffset,omitempty" valid:"omitempty,timezoneoffset"`
+	ConversionOffset *int          `json:"conversionOffset,omitempty" bson:"conversionOffset,omitempty" valid:"omitempty,required"`
+	ClockDriftOffset *int          `json:"clockDriftOffset,omitempty" bson:"clockDriftOffset,omitempty" valid:"omitempty,required"`
+	Payload          interface{}   `json:"payload,omitempty" bson:"payload,omitempty" valid:"omitempty,payload"`
+	Annotations      []interface{} `json:"annotations,omitempty" bson:"annotations,omitempty" valid:"omitempty,annotations"`
+
+	//used for versioning and de-deping
+	Internal `bson:",inline"`
+}
+
+//Internal are existing fields used for versioning and de-deping
+type Internal struct {
+	CreatedTime   string `json:"createdTime" bson:"createdTime" valid:"timestr"`
+	GroupID       string `json:"-" bson:"_groupId" valid:"required"`
+	ActiveFlag    bool   `json:"-" bson:"_active" valid:"required"`
+	SchemaVersion int    `json:"-" bson:"_schemaVersion" valid:"required,min=0"`
+	Version       int    `json:"-" bson:"_version,omitempty" valid:"-"`
+}
+
+var (
+	//InternalFields are what we only use internally in the service and don't wish to return to any clients
+	InternalFields = map[string]interface{}{
+		"_groupId":       0,
+		"_active":        0,
+		"_schemaVersion": 0,
+		"_version":       0,
+		"createdTime":    0,
+	}
+)
+
+const (
+	UserIDField  = "userId"
+	GroupIDField = "groupId"
+
+	deviceIDField = "deviceId"
+	uploadIDField = "uploadId"
+
+	timezoneOffsetField   = "timezoneOffset"
+	conversionOffsetField = "conversionOffset"
+	clockDriftOffsetField = "clockDriftOffset"
+
+	typeField       = "type"
+	timeField       = "time"
+	deviceTimeField = "deviceTime"
+
+	payloadField     = "payload"
+	annotationsField = "annotations"
+
+	tzValidationLowerLimit = -840
+	tzValidationUpperLimit = 720
+	timeStrValidationMsg   = "Times need to be ISO 8601 format and not in the future"
+
+	timeStringTag     validate.ValidationTag = "timestr"
+	timeObjectTag     validate.ValidationTag = "timeobject"
+	timeZoneOffsetTag validate.ValidationTag = "timezoneoffset"
+	payloadTag        validate.ValidationTag = "payload"
+	annotationsTag    validate.ValidationTag = "annotations"
+)
+
+func BuildBase(datum Datum, errs validate.ErrorProcessing) Base {
+
+	base := Base{
+		_ID:              bson.NewObjectId(),
+		ID:               bson.NewObjectId().Hex(),
+		UserID:           ToString(UserIDField, datum[UserIDField], errs),
+		DeviceID:         ToString(deviceIDField, datum[deviceIDField], errs),
+		UploadID:         ToString(uploadIDField, datum[uploadIDField], errs),
+		Time:             ToString(timeField, datum[timeField], errs),
+		Type:             ToString(typeField, datum[typeField], errs),
+		Payload:          datum[payloadField],
```

Comment:
Probably want some sort of TODO to provide a minimum validation for the payload field.